runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -639,6 +639,165 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
+               J9Class* defcClazz = (J9Class*)TR::Compiler->cls.classFromJavaLangClass(comp(), defc);
+               if (defcClazz->initializeStatus == J9ClassInitSucceeded)
+                  {
+                  removeCall = true;
+                  if (trace())
+                     {
+                     traceMsg(comp(), "defc in MethodHandle %p has been initialized\n", mh);
+                     }
+                  }
+               }
+
+            if (removeCall
+                && performTransformation(comp(), "%sRemoving call node %s [" POINTER_PRINTF_FORMAT "]\n", OPT_DETAILS, node->getOpCode().getName(), node))
+               {
+               TR::Node* receiverChild = node->getChild(0);
+               TR::TransformUtil::transformCallNodeToPassThrough(this, node, _curTree, receiverChild);
+               TR::DebugCounter::incStaticDebugCounter(comp(), TR::DebugCounter::debugCounterName(comp(), "constrainCall/(%s)", signature));
+               return;
+               }
+            break;
+            }
+         case TR::java_lang_invoke_DirectHandle_nullCheckIfRequired:
+            {
+            TR::Node* mh = node->getChild(0);
+            TR::Node* receiver = node->getChild(1);
+            bool mhConstraintGlobal, receiverConstraintGlobal;
+            TR::VPConstraint* mhConstraint = getConstraint(mh, mhConstraintGlobal);
+            TR::VPConstraint* receiverConstraint = getConstraint(receiver, receiverConstraintGlobal);
+            bool removeCall = false;
+            if (receiverConstraint
+                && receiverConstraint->isNonNullObject())
+               {
+               removeCall = true;
+               if (trace())
+                  {
+                  traceMsg(comp(), "receiver child %p for DirectHandle.nullCheckIfRequired is non-null\n", receiver);
+                  }
+               }
+            else if (mhConstraint
+                     && mhConstraint->getKnownObject()
+                     && mhConstraint->isNonNullObject()
+                     && mhConstraint->getClass())
+               {
+               TR_OpaqueClassBlock* mhClass = mhConstraint->getClass();
+               int32_t modifierOffset = comp()->fej9()->getInstanceFieldOffset(mhClass, "final_modifiers", 15, "I", 1);
+               if (modifierOffset < 0)
+                  break;
+               uintptrj_t* mhLocation = getObjectLocationFromConstraint(mhConstraint);
+
+               TR::VMAccessCriticalSection nullCheckIfRequired(comp(),
+                                                        TR::VMAccessCriticalSection::tryToAcquireVMAccess);
+               if (!nullCheckIfRequired.hasVMAccess())
+                  break;
+               uintptrj_t mhObject = comp()->fej9()->getStaticReferenceFieldAtAddress((uintptrj_t)mhLocation);
+               int32_t modifier = comp()->fej9()->getInt32FieldAt(mhObject, modifierOffset);
```

Comment:
This access should be safe because a valid `modifierOffset` was found from the class of `mhObject` :+1: 

Here, if the code is on a dead path, that class might not be `DirectHandle`, but it contains an integer-typed field at `modifierOffset`, and it doesn't matter whether or not the call is removed.